src/api/authentication.js

```diff
@@ -56,57 +83,103 @@ export async function authenticate (ctx, next) {
   const from = new Date()
   from.setSeconds(from.getSeconds() - authWindowSeconds)
 
-  if ((requestDate < from) || (requestDate > to)) {
+  if (requestDate < from || requestDate > to) {
     // request expired
-    logger.info(`API request made by ${email} from ${ctx.request.host} has expired, denying access`)
-    ctx.status = 401
-    auditAuthFailure()
-    return
+    ctx.throw(
+      401,
+      `API request made by ${email} from ${ctx.request.host} has expired, denying access`,
+      { email }
+    )
   }
-  const user = await UserModelAPI.findOne({ email: caseInsensitiveRegex(email) })
-  ctx.authenticated = user
 
-  if (!user) {
+  const user = await UserModelAPI.findOne({ email: caseInsensitiveRegex(email) })
+  if (user == null) {
     // not authenticated - user not found
-    logger.info(`No user exists for ${email}, denying access to API, request originated from ${ctx.request.host}`)
-    ctx.status = 401
-    auditAuthFailure()
-    return
+    ctx.throw(
+      401,
+      `No user exists for ${email}, denying access to API, request originated from ${ctx.request.host}`,
+      { email }
+    )
   }
 
   const hash = crypto.createHash('sha512')
   hash.update(user.passwordHash)
   hash.update(authSalt)
   hash.update(authTS)
 
-  if (authToken === hash.digest('hex')) {
-    // authenticated
+  if (authToken !== hash.digest('hex')) {
+    // not authenticated - token mismatch
+    ctx.throw(
+      401,
+      `API token did not match expected value, denying access to API, the request was made by ${email} from ${ctx.request.host}`,
+      { email }
+    )
+  }
 
-    if (ctx.path === '/transactions') {
-      if (!ctx.query.filterRepresentation || (ctx.query.filterRepresentation !== 'full')) {
-        // exempt from auditing success
-        await next()
-        return
-      }
-    } else {
-      for (const pathTest of Array.from(auditingExemptPaths)) {
-        if (pathTest.test(ctx.path)) {
-          // exempt from auditing success
-          await next()
-          return
-        }
-      }
+  return user
+}
+
+async function authenticateRequest (ctx) {
+  // First attempt basic authentication
+  let user = await authenticateBasic(ctx)
+  if (user == null) {
+    // Otherwise try token based authentication
+    user = await authenticateToken(ctx)
```

Comment:
I feel like we should maybe add a config option to enable both auth methods so one can be disabled in particular circumstances. Like if running over raw http or if some of the features (like replay prevention) from token auth are desired.

Reply:
The API is only available over HTTPS so prevention of replay attacks is handled by TLS. A config sounds like a good idea though. I will add one in and default to using the token style auth only so that the behaviour does not change.